server/src/main/java/org/apache/druid/server/QueryResource.java

```diff
@@ -252,24 +252,30 @@ public void write(OutputStream outputStream) throws WebApplicationException
             )
             .header("X-Druid-Query-Id", queryId);
 
-        if (responseContext.get(ResponseContext.CTX_ETAG) != null) {
-          builder.header(HEADER_ETAG, responseContext.get(ResponseContext.CTX_ETAG));
-          responseContext.remove(ResponseContext.CTX_ETAG);
+        if (responseContext.get(ResponseContext.Key.ETAG) != null) {
+          builder.header(HEADER_ETAG, responseContext.get(ResponseContext.Key.ETAG));
+          responseContext.remove(ResponseContext.Key.ETAG);
         }
 
         DirectDruidClient.removeMagicResponseContextFields(responseContext);
 
         //Limit the response-context header, see https://github.com/apache/incubator-druid/issues/2331
         //Note that Response.ResponseBuilder.header(String key,Object value).build() calls value.toString()
         //and encodes the string using ASCII, so 1 char is = 1 byte
-        String responseCtxString = responseContext.serializeWith(jsonMapper);
-        if (responseCtxString.length() > RESPONSE_CTX_HEADER_LEN_LIMIT) {
-          log.warn("Response Context truncated for id [%s] . Full context is [%s].", queryId, responseCtxString);
-          responseCtxString = responseCtxString.substring(0, RESPONSE_CTX_HEADER_LEN_LIMIT);
+        final ResponseContext.SerializationResult serializationResult = responseContext.serializeWith(
+            jsonMapper,
+            RESPONSE_CTX_HEADER_LEN_LIMIT
+        );
+        if (serializationResult.isReduced()) {
+          log.warn(
```

Comment:
Should Druid cluster operators monitor these messages? Can they do anything about them? If not, this should probably be `info()`. See #7362.

Reply:
I'm not sure about that, even left the log message as is although the context is not truncated anymore but "reduced". According to the [corresponding PR discussion](https://github.com/apache/incubator-druid/pull/2336), it was important to have a log message with full context, it's likely someone has a filter in a log aggregator for this kind of message.
BTW I see your point and started discussion to only mention backward compatibility for log filters.

Reply:
Please add a comment like "Whether or not this logging statement should properly be on the WARN level (which is unclear), it's kept on the warn level for backward compatibility: see https://github.com/apache/incubator-druid/pull/2336".

(If I understood your comment correctly.)

Reply:
Since the change will be tagged as `incompatible` I decided to update the log level to `info`